src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala

```diff
@@ -0,0 +1,228 @@
+        submissionRec <- subRecs.headOption
+
+        //The workspace
+        wsRecs <- dataAccess.workspaceQuery.findByIdQuery(submissionRec.workspaceId).result
+        workspaceRec <- wsRecs.headOption
+
+        //The workspace's billing project
+        bpOpt <- dataAccess.rawlsBillingProjectQuery.load(RawlsBillingProjectName(workspaceRec.namespace))
+        billingProject <- bpOpt
+
+        //The person who submitted the submission, and their token
+        submitterOpt <- dataAccess.rawlsUserQuery.load(RawlsUserRef(RawlsUserSubjectId(submissionRec.submitterId)))
+        submitter <- submitterOpt
+        tokenOpt <- DBIO.from(googleServicesDAO.getToken(submitter))
+        token <- tokenOpt
+
+        //The wdl
+        mcOpt <- dataAccess.methodConfigurationQuery.loadMethodConfigurationById(submissionRec.methodConfigurationId)
+        methodConfig <- mcOpt
+        wdl <- getWdl(methodConfig)
+      } yield {
+
+        val wfOpts = buildWorkflowOpts(workspaceRec, submissionId, submitter, token, billingProject)
+
+        val wfInputsBatch = unpackedWfs map { wf =>
+          val methodProps = wf.inputResolutions map {
+            case svv: SubmissionValidationValue if svv.value.isDefined =>
+              svv.inputName -> svv.value.get
+          }
+          MethodConfigResolver.propertiesToWdlInputs(methodProps.toMap)
+        }
+
+        //yield the things we're going to submit to Cromwell
+        (wdl, wfRecs.map(_.id), wfInputsBatch, wfOpts)
+      }
+    }
+
+    val cromwellSubmission = for {
+      (wdl: String, workflowRecIds: Seq[Long], wfInputsBatch: Seq[String], wfOpts: Option[String]) <- workflowBatchFuture
+      workflowSubmitResult <- executionServiceDAO.submitWorkflows(wdl, wfInputsBatch, Option(wfOpts.toJson.toString), getUserInfo(credential))
+    } yield {
+      workflowRecIds.zip(workflowSubmitResult)
+    }
+
+    cromwellSubmission flatMap { results =>
+      dataSource.inTransaction { dataAccess =>
+        //save successes as submitted workflows and hook up their cromwell ids
+        val successes = results collect {
+          case (wfRecId, Left(success: ExecutionServiceStatus)) =>
+            //TODO: batchify this update?
+            dataAccess.workflowQuery.findWorkflowById(wfRecId).map(u => u.externalid).update(success.id)
+            wfRecId
+        }
+        dataAccess.workflowQuery.batchUpdateStatus(successes, WorkflowStatuses.Submitted)
+
+        //save error messages into failures and flip them to Failed
+        val failures = results collect {
+          case (wfRecId, Right(failure: ExecutionServiceFailure)) =>
+            //TODO: batchify this saveMessages?
+            dataAccess.workflowQuery.saveMessages(Seq(AttributeString(failure.message)), wfRecId)
+            wfRecId
+        }
+        dataAccess.workflowQuery.batchUpdateStatus(failures, WorkflowStatuses.Failed)
+      } map { _ => LookForWorkflows }
+    } recover {
```

Comment:
Is this the right place for this recover? I was toying with putting them on each intermediate future (i.e. one on `workflowBatchFuture` and one on `cromwellSubmission` but then I got confused by whether it'd pass the failures along and recover multiple times for the same failure.